runtime/lua/vim/lsp/util.lua

```diff
@@ -612,15 +612,77 @@ function M.text_document_completion_list_to_complete_items(result, prefix)
   return matches
 end
 
+
+--- Rename old_fname to new_fname
+--
+--@param opts (table)
+--         overwrite? bool
+--         ignoreIfExists? bool
+function M.rename(old_fname, new_fname, opts)
+  opts = opts or {}
+  local bufnr = vim.fn.bufadd(old_fname)
+  vim.fn.bufload(bufnr)
+  local target_exists = vim.loop.fs_stat(new_fname) ~= nil
+  if target_exists and not opts.overwrite or opts.ignoreIfExists then
+    return
+  end
+  local ok, err = os.rename(old_fname, new_fname)
+  assert(ok, err)
+  api.nvim_buf_call(bufnr, function()
+    vim.cmd('saveas! ' .. vim.fn.fnameescape(new_fname))
+  end)
+end
+
+
+local function create_file(change)
+  local opts = change.options or {}
+  -- from spec: Overwrite wins over `ignoreIfExists`
+  local fname = vim.uri_to_fname(change.uri)
+  if not opts.ignoreIfExists or opts.overwrite then
+    local file = io.open(fname, 'w')
+    file:close()
+  end
+  vim.fn.bufadd(fname)
```

Comment:
is it weird that createfile also adds a buffer?

Reply:
Oh nvm, not in this context

Reply:
I actually just hit on this and was digging to see if there was a conversation about it. I just hit on a scenario where this sort of screws up the intended flow of some edits, and I'm curious if this should actually be here. In Metals you can create an empty file just with a package declaration, and this can be done with a request like this:

```
[Trace - 01:07:28 PM] Sending request 'workspace/applyEdit - (3)'
Params: {
  "edit": {
    "changes": {},
    "documentChanges": [
      {
        "uri": "file:///Users/ckipp/Documents/scala-workspace/sanity/Sanity/src/example/Foo.scala",
        "kind": "create"
      },
      {
        "textDocument": {
          "version": 1,
          "uri": "file:///Users/ckipp/Documents/scala-workspace/sanity/Sanity/src/example/Foo.scala"
        },
        "edits": [
          {
            "range": {
              "start": {
                "line": 0,
                "character": 0
              },
              "end": {
                "line": 0,
                "character": 0
              }
            },
            "newText": "package example\n\n\n"
          }
        ]
      }
    ]
  }
}
```

The first creates the file, and the second adds the text. Metals also has some extra logic that if you do a `didOpen` on a new Scala file that doesn't have a package declaration it will infer it from the context and insert it. Because of the way this is processed in the loop, before the second edit happens the `bufadd` call results in a `didOpen` with file that is empty:

```
[Trace - 01:07:28 PM] Received notification 'textDocument/didOpen'
Params: {
  "textDocument": {
    "uri": "file:///Users/ckipp/Documents/scala-workspace/sanity/Sanity/src/example/Foo.scala",
    "languageId": "scala",
    "version": 0,
    "text": "\n"
  }
}
```
Then Metals will return a:

```
[Trace - 01:07:28 PM] Sending request 'workspace/applyEdit - (4)'
Params: {
  "edit": {
    "changes": {
      "file:///Users/ckipp/Documents/scala-workspace/sanity/Sanity/src/example/Foo.scala": [
        {
          "range": {
            "start": {
              "line": 0,
              "character": 0
            },
            "end": {
              "line": 0,
              "character": 0
            }
          },
          "newText": "package example\n\n"
        }
      ]
    }
  }
}
```
Which you guess it will result in your file being:

```
package example

package example
```

I'm curious how we can avoid this, and looking at the spec, I don't see any reference to a `CreateFile` event being actually opened, and I think this is a bit of an assumption being made here that it should be. Do you think this can be removed? Maybe you have some more context about why it was included @mfussenegger?

Reply:
I don't remember if I hit issues without it. But from a user perspective I'd somehow assume that the buffer should be added if created via LSP.

One thing we could do is defer opening created files until after all edits have been processed.